File: sensei/testing/testProgrammableDataAdaptor.cpp
#include "ProgrammableDataAdaptor.h"
#include "MeshMetadata.h"
#include "Histogram.h"

#include <svtkDataObject.h>
#include <svtkImageData.h>
#include <svtkPointData.h>
#include <svtkDoubleArray.h>

#include <vector>
#include <iostream>

#include <mpi.h>

using std::cerr;
using std::endl;

int main(int argc, char **argv)
{
  MPI_Init(&argc, &argv);

  //
  //                                              *
  //                                              * *
  //                                            * * *
  //                                            * * * *
  //                                          * * * * *
  //                                        * * * * * * *
  std::vector<unsigned int> baselineHist = {1,2,4,6,5,3,1};
  //                                        0,1,2,3,4,5,6

  std::vector<double> data = {0, 1,1, 2,2,2,2,
    3,3,3,3,3,3, 4,4,4,4,4, 5,5,5, 6};

  // get number of meshes callback
  auto getNumberOfMeshes = [](unsigned int &n) -> int
    {
    cerr << "===getNumberOfMeshes" << endl;
    n = 1;
    return 0;
    };

  // get mesh name callback
  auto getMeshMetadata = [](unsigned int id, sensei::MeshMetadataPtr &metadata) -> int
    {
    cerr << "===getMeshMetadata" << endl;
    if (id == 0)
      {
      metadata->MeshName = "image";
      metadata->MeshType = SVTK_IMAGE_DATA;
      metadata->BlockType = SVTK_IMAGE_DATA;
      metadata->NumBlocks = 1;
      metadata->NumBlocksLocal = {1};
      metadata->NumArrays = 1;
      metadata->ArrayName = {"data"};
      metadata->ArrayCentering = {svtkDataObject::POINT};
      metadata->ArrayType = {SVTK_DOUBLE};
      metadata->ArrayComponents = {1};
      return 0;
      }
    return -1;
    };

  // get mesh callback
  auto getMesh = [&data](const std::string &meshName,
    bool, svtkDataObject *&mesh) -> int
    {
    cerr << "===getMesh" << endl;
    if (meshName == "image")
      {
      svtkImageData *im = svtkImageData::New();
      im->SetDimensions(data.size(), 1, 1);
      mesh = im;
      return 0;
      }
    return -1;
    };

  // add array callback
  auto addArray = [&data](svtkDataObject *mesh,
    const std::string &meshName, int assoc, const std::string &name) -> int
    {
    cerr << "===addArray" << endl;
    if ((meshName == "image") && (assoc == svtkDataObject::POINT) && (name == "data"))
      {
      svtkDoubleArray *da = svtkDoubleArray::New();
      da->SetName("data");
      da->SetArray(data.data(), data.size(), 1);

      static_cast<svtkImageData*>(mesh)->GetPointData()->AddArray(da);
      da->Delete();
      return 0;
      }
    return -1;
    };

  // release data callback
  auto releaseData = []() -> int
    {
    cerr << "===releaseData" << endl;
    return 0;
    };

  sensei::ProgrammableDataAdaptor *pda = sensei::ProgrammableDataAdaptor::New();
  pda->SetGetNumberOfMeshesCallback(getNumberOfMeshes);
  pda->SetGetMeshMetadataCallback(getMeshMetadata);
  pda->SetGetMeshCallback(getMesh);
  pda->SetAddArrayCallback(addArray);
  pda->SetReleaseDataCallback(releaseData);


  sensei::MeshMetadataPtr mmd = sensei::MeshMetadata::New();
  pda->GetMeshMetadata(0, mmd);

  sensei::Histogram *ha = sensei::Histogram::New();
  ha->Initialize(7, mmd->MeshName, mmd->ArrayCentering[0], mmd->ArrayName[0], "");
  ha->Execute(pda);

  pda->ReleaseData();

  sensei::Histogram::Data result;
  ha->GetHistogram(result);

  int status = -1;
  if (result.Histogram == baselineHist)
    status = 0;

  pda->Delete();
  ha->Delete();

  MPI_Finalize();

  return status;
}

File: sensei/testing/testProgrammableDataAdaptor.py
from mpi4py import *
import sys
import sensei
import svtk, svtk.numpy_support as svtknp
import numpy as np

#                     *
#                     * *
#                   * * *
#                   * * * *
#                 * * * * *
#               * * * * * * *
baselineHist = [1,2,4,6,5,3,1]
#               0,1,2,3,4,5,6

data = np.array([0, 1,1, 2,2,2,2, \
  3,3,3,3,3,3, 4,4,4,4,4, \
  5,5,5, 6], dtype=np.float64)

# get num meshes callback
def getNumMeshes():
  sys.stderr.write('===getNumMeshes\n')
  return 1

def getMeshMetadata(i, flags):
  sys.stderr.write('===getMeshMetadata\n')
  sys.stderr.write('flags=%s'%(flags))
  if i == 0:
    md = sensei.MeshMetadata.New()
    md.MeshName = 'image'
    md.MeshType = svtk.SVTK_IMAGE_DATA
    md.BlockType = svtk.SVTK_IMAGE_DATA
    md.NumArrays = 1
    md.ArrayName = ['data']
    md.ArrayCentering = [svtk.svtkDataObject.POINT]
    md.ArrayComponents = [1]
    md.ArrayType = [svtk.SVTK_DOUBLE]
    return md
  raise RunTimeError('no mesh %d'%(i))

# get mesh callback
def getMesh(meshName, structureOnly):
  sys.stderr.write('===getMesh\n')
  if (meshName == 'image'):
    im = svtk.svtkImageData()
    im.SetDimensions(len(data), 1, 1)
    return im
  raise RuntimeError('failed to get mesh')

# add array callback
def addArray(mesh, meshName, assoc, arrayName):
  sys.stderr.write('===addArray\n')
  if ((meshName == 'image') and (assoc == svtk.svtkDataObject.POINT) \
    and (arrayName == 'data')):
    da = svtknp.numpy_to_svtk(data)
    da.SetName('data')
    mesh.GetPointData().AddArray(da)
    return
  raise RuntimeError('failed to add array')

# release data callback
def releaseData():
  sys.stderr.write('===releaseData\n')

m = getMesh('image', False)
addArray(m, 'image', svtk.svtkDataObject.POINT, 'data')
sys.stderr.write('m = %s\n'%str(m))
sys.exit(0)



pda = sensei.ProgrammableDataAdaptor.New()
pda.SetGetNumberOfMeshesCallback(getNumMeshes)
pda.SetGetMeshMetadataCallback(getMeshMetadata)
pda.SetGetMeshCallback(getMesh)
pda.SetAddArrayCallback(addArray)
pda.SetReleaseDataCallback(releaseData)

mmd = pda.GetMeshMetadata(0)
sys.stderr.write('mmd = %s\n'%str(mmd))

meshName = mmd.MeshName
arrayName = mmd.ArrayName[0]
arrayCen = mmd.ArrayCentering[0]

ha = sensei.Histogram.New()
ha.Initialize(7, meshName, arrayCen, arrayName, '')
ha.Execute(pda)

hmin,hmax,hist = ha.GetHistogram()

result = -1
if hist == baselineHist:
  result = 0

ha.Delete()

pda.ReleaseData()
pda.Delete()

sys.exit(result)
